zerver/tests/test_auth_backends.py

```diff
@@ -3816,6 +3816,55 @@ def test_ldap_auth_email_auth_organization_restriction_success(self) -> None:
         )
         self.assert_json_success(result)
 
+    @override_settings(
+        AUTHENTICATION_BACKENDS=("zproject.backends.ZulipLDAPAuthBackend",),
+        AUTH_LDAP_USER_ATTR_MAP={"full_name": "cn"},
+        AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL={
+            "zulip": {"test1": "test", "test2": "testing"},
+            "anotherRealm": {"test2": "test2"},
+        },
+    )
+    def test_ldap_auth_email_auth_advanced_organization_restriction(self) -> None:
+        self.init_default_ldap_database()
+        # We do test two combinations here:
+        # The first user has no attribute set
+        result = self.client_post(
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_error(result, "Your username or password is incorrect.", 403)
+        # Using test2 here because its the second in dict
+        self.change_ldap_user_attr("hamlet", "test2", "testing")
+        result = self.client_post(
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_success(result)
+        # Setting test2 to wrong value
+        self.change_ldap_user_attr("hamlet", "test2", "test")
+        result = self.client_post(
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_error(result, "Your username or password is incorrect.", 403)
+
+    @override_settings(
+        AUTHENTICATION_BACKENDS=("zproject.backends.ZulipLDAPAuthBackend",),
+        AUTH_LDAP_USER_ATTR_MAP={"full_name": "cn"},
+        AUTH_LDAP_ADVANCED_REALM_ACCESS_CONTROL={"notZulip": {"department": "zulip"}},
+    )
+    def test_ldap_auth_email_auth_advanced_organization_restriction_wrong_configuration(
+        self,
+    ) -> None:
+        self.init_default_ldap_database()
+        # A wrong config/not existing realm should not grant any access
+        self.change_ldap_user_attr("hamlet", "department", "zulip")
+        result = self.client_post(
+            "/api/v1/fetch_api_key",
+            dict(username=self.example_email("hamlet"), password=self.ldap_password("hamlet")),
+        )
+        self.assert_json_error(result, "Your username or password is incorrect.", 403)
```

Comment:
I think my comments on the test setup/structure for the first commit apply to this second commit as well.